tmt/steps/provision/artemis.py

```diff
@@ -0,0 +1,631 @@
+            :py:mod:`requests` library.
+        """
+
+        request_kwargs = request_kwargs or {}
+
+        if params:
+            request_kwargs['params'] = params
+
+        return self.query(path, request_kwargs=request_kwargs)
+
+    def delete(
+            self,
+            path: str,
+            request_kwargs: Optional[Dict[str, Any]] = None
+            ) -> requests.Response:
+        """
+        Delete - or request removal of - a resource.
+
+        :param path: API path to contact.
+        :param request_kwargs: optional request options, as supported by
+            :py:mod:`requests` library.
+        """
+
+        return self.query(path, method='delete', request_kwargs=request_kwargs)
+
+
+class ProvisionArtemis(tmt.steps.provision.ProvisionPlugin):
+    """
+    Provision guest using Artemis backend
+
+    Minimal configuration could look like this:
+
+        provision:
+            how: artemis
+            image: Fedora
+            api-url: https://your-artemis.com/
+
+    Note that the actual value of "image" depends on what images - or "composes"
+    as Artemis calls them - supports and can deliver.
+
+    Note that "api-url" can be also given via ARTEMIS_API_URL environment
+    variable.
+
+    Full configuration example:
+
+        provision:
+            how: artemis
+
+            # Artemis API
+            api-url: https://your-artemis.com/
+            api-version: 0.0.32
+
+            # Mandatory environment properties
+            image: Fedora
+
+            # Optional environment properties
+            arch: aarch64
+            pool: optional-pool-name
+
+            # Provisioning process control (optional)
+            priority-group: custom-priority-group
+            keyname: custom-SSH-key-name
+
+            # Labels to be attached to guest request (optional)
+            user-data:
+              foo: bar
+
+            # Timeouts and deadlines (optional)
+            provision-timeout: 3600
+            provision-tick: 10
+            api-timeout: 600
+            api-retries: 5
+            api-retry-backoff-factor: 1
+    """
+
+    # Guest instance
+    _guest = None
+
+    # Supported methods
+    _methods = [
+        tmt.steps.Method(name='artemis', doc=__doc__, order=50),
+        ]
+
+    _keys = [
+        'api-url',
+        'api-version',
+        'arch',
+        'image',
+        'hardware',
+        'pool',
+        'priority-group',
+        'keyname',
+        'user-data',
+        'provision-timeout',
+        'provision-tick',
+        'api-timeout',
+        'api-retries',
+        'api-retry-backoff-factor']
+
+    @classmethod
+    def options(cls, how=None):
+        """ Prepare command line options for Artemis """
+        return [
+            click.option(
+                '--api-url', metavar='URL',
+                help="Artemis API URL.",
+                envvar='ARTEMIS_API_URL',
+                required=True
```

Comment:
Should it be `required` if the value can also be defined in the `provision` step config?

Reply:
Ah, I thought `required` here applies to the option in general. If it's just for *CLI* option, then no, it's not necessary to use `required` here.

Reply:
Fixed.